cmds/upspin_sos/service.go

```diff
@@ -78,9 +78,23 @@ func (us UpspinService) setFileData(path string) error {
 }
 
 func (us UpspinService) setKeys(path string) error {
-	// execute as user. This command generates files with elevated permissions otherwise
-	keygen := exec.Command("upspin", "keygen", fmt.Sprintf("-secretseed=%v", us.Seed), path)
-	err := keygen.Run()
+	// check if keys are set already
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	_, err = f.Readdir(1)
+	args := []string{"keygen", fmt.Sprintf("-secretseed=%v", us.Seed)}
+	// if the directory is populated, rotate the keys instead of generating new ones
+	// this method of appending args is ugly, but they have to be in this specific order:
+	// upspin keygen <-rotate> -secretseed=<seed> path
+	if err == nil {
+		args = append(args, "-rotate")
+	}
+
+	keygen := exec.Command("upspin", append(args, path)...)
```

Comment:
or lines 96-101 could simply be
return exec.Command("upspin", append(args, path)...).Run()
replace all those lines with 1 line.